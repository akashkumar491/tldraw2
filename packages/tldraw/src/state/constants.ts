export const FIT_TO_SCREEN_PADDING = 128
