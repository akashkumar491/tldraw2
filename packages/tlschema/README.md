# @tldraw/tlschema

This package houses type definitions, schema migrations, and other type metadata for any persisted data in tldraw.

There are three main kinds of types:

- Record types

  These are root record types added to the `Store` class. They are defined in the `./src/records` directory.

- Shape types

  These are subtypes of the root TLShape record type. They allow specifying a unique name and custom props for a particular kind of shape.

- Asset types

  These are subtypes of the root TLAsset record type. They allow specifying a unique name and custom props for a particular kind of asset.

## Creating a new record type

To create a new record type called, e.g. `TLBanana`, run

```sh
# Create the new file in ./records/TLBanana.ts
yarn new-record TLBanana
# Rebuild the index files to make sure it is exported
yarn index
```

You may then customize your new blank `TLBanana` type!

## Creating a new shape type

To create a new shape type called, e.g. `TLMapboxShape`, run

```sh
# Create the new file in ./shapes/TLMapboxShape.ts
yarn new-shape TLMapboxShape
# Rebuild the index files to make sure it is exported
yarn index
```

You may then customize your new blank `TLMapboxShape` type!

## Creating a new asset type

To create a new shape type called, e.g. `TLZipFileAsset`, run

```sh
# Create the new file in ./assets/TLZipFileAsset.ts
yarn new-asset TLZipFileAsset
# Rebuild the index files to make sure it is exported
yarn index
```

You may then customize your new blank `TLZipFileAsset` type!

# Adding migrations

If you make any kind of change to any persisted data shape in this package, you must add migrations that are able
to convert old versions to new versions, and vice-versa.

If you are making a change that affects the structure of a record, shape, or asset, update the migrations in the same file as the record, shape, or asset is defined.

If you are making a change that affects the structure of the store (e.g. renaming or deleting a type, consolidating two shape types into one, etc), add your changes in the migrations in `schema.ts`.

After making your changes, add a new version number, using a meaninful name. For example, if you add a new property
to the `TLShape` type called `ownerId` that points to a user, you might do this:

In `TLShape.ts`

```diff
 const Versions = {
   Initial: 1,
+  AddOwnerId: 2,
 } as const
```

and then in the `TLShape` type

```diff
   x: number
   y: number
+  ownerId: ID<TLUser> | null
   props: Props
   parentId: ID<TLShape> | ID<TLPage>
```

and then adding a migration:

```diff
 export const shapeTypeMigrations = defineMigrations({
   // STEP 2: Update the current version to point to your latest version
   currentVersion: Versions.Initial,
   firstVersion: Versions.Initial,
   migrators: {
     // STEP 3: Add an up+down migration for the new version here
+    [Versions.AddOwnerId]: {
+      // add ownerId property
+      up: (shape) => ({...shape, ownerId: null}),
+      // remove ownerId property
+      down: ({ownerId, ...shape}) => shape,
+    }
   },
```

After you've added your migration, make sure to add a test for it in `src/migrations.test.ts`. It will complain if you do not!

## License

The source code in this repository (as well as our 2.0+ distributions and releases) are currently licensed under Apache-2.0. These licenses are subject to change in our upcoming 2.0 release. If you are planning to use tldraw in a commercial product, please reach out at [hello@tldraw.com](mailto://hello@tldraw.com).
