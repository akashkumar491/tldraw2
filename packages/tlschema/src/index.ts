export { type TLRecord } from './TLRecord'
export {
	createIntegrityChecker,
	onValidationFailure,
	type TLStore,
	type TLStoreProps,
	type TLStoreSchema,
	type TLStoreSnapshot,
} from './TLStore'
export { createAssetValidator, type TLBaseAsset } from './assets/TLBaseAsset'
export {
	bookmarkAssetMigrations,
	bookmarkAssetTypeValidator,
	type TLBookmarkAsset,
} from './assets/TLBookmarkAsset'
export {
	imageAssetMigrations,
	imageAssetTypeValidator,
	type TLImageAsset,
} from './assets/TLImageAsset'
export {
	videoAssetMigrations,
	videoAssetTypeValidator,
	type TLVideoAsset,
} from './assets/TLVideoAsset'
export { createPresenceStateDerivation } from './createPresenceStateDerivation'
export { createTLSchema, type SchemaShapeInfo } from './createTLSchema'
export { CLIENT_FIXUP_SCRIPT, fixupRecord } from './fixup'
export { TL_COLOR_TYPES, uiColorTypeValidator, type TLColor } from './misc/TLColor'
export {
	TL_CURSOR_TYPES,
	cursorTypeValidator,
	cursorValidator,
	type TLCursor,
	type TLCursorType,
} from './misc/TLCursor'
export {
	TL_HANDLE_TYPES,
	handleTypeValidator,
	type TLHandle,
	type TLHandlePartial,
	type TLHandleType,
} from './misc/TLHandle'
export { TL_SCRIBBLE_STATES, scribbleTypeValidator, type TLScribble } from './misc/TLScribble'
export { type Box2dModel, type Vec2dModel } from './misc/geometry-types'
export { idValidator } from './misc/id-validator'
export {
	AssetRecordType,
	assetIdValidator,
	assetTypeMigrations,
	assetTypeValidator,
	type TLAsset,
	type TLAssetId,
	type TLAssetPartial,
	type TLAssetShape,
} from './records/TLAsset'
export {
	CameraRecordType,
	cameraTypeValidator,
	type TLCamera,
	type TLCameraId,
} from './records/TLCamera'
export {
	DocumentRecordType,
	TLDOCUMENT_ID,
	documentTypeValidator,
	type TLDocument,
} from './records/TLDocument'
export {
	InstanceRecordType,
	instanceIdValidator,
	instanceTypeMigrations,
	instanceTypeValidator,
	type TLInstance,
	type TLInstanceId,
	type TLInstancePropsForNextShape,
} from './records/TLInstance'
export {
	PageRecordType,
	isPageId,
	pageIdValidator,
	pageTypeValidator,
	type TLPage,
	type TLPageId,
} from './records/TLPage'
export {
	InstancePageStateRecordType,
	instancePageStateMigrations,
	instancePageStateTypeValidator,
	type TLInstancePageState,
	type TLInstancePageStateId,
} from './records/TLPageState'
export {
	PointerRecordType,
	TLPOINTER_ID,
	pointerTypeValidator,
	type TLPointer,
	type TLPointerId,
} from './records/TLPointer'
export { InstancePresenceRecordType, type TLInstancePresence } from './records/TLPresence'
export {
	createShapeId,
	isShape,
	isShapeId,
	rootShapeTypeMigrations,
	type TLDefaultShape,
	type TLNullableShapeProps,
	type TLParentId,
	type TLShape,
	type TLShapeId,
	type TLShapePartial,
	type TLShapeProp,
	type TLShapeProps,
	type TLUnknownShape,
} from './records/TLShape'
export {
	UserDocumentRecordType,
	userDocumentTypeMigrations,
	userDocumentTypeValidator,
	type TLUserDocument,
	type TLUserDocumentId,
} from './records/TLUserDocument'
export {
	TL_ARROW_TERMINAL_TYPE,
	arrowShapeTypeMigrations,
	arrowShapeTypeValidator,
	arrowTerminalTypeValidator,
	type TLArrowHeadModel,
	type TLArrowShape,
	type TLArrowShapeProps,
	type TLArrowTerminal,
	type TLArrowTerminalType,
} from './shapes/TLArrowShape'
export {
	createShapeValidator,
	parentIdValidator,
	shapeIdValidator,
	type TLBaseShape,
} from './shapes/TLBaseShape'
export {
	bookmarkShapeTypeMigrations,
	bookmarkShapeTypeValidator,
	type TLBookmarkShape,
	type TLBookmarkShapeProps,
} from './shapes/TLBookmarkShape'
export {
	TL_DRAW_SHAPE_SEGMENT_TYPE,
	drawShapeTypeMigrations,
	drawShapeTypeValidator,
	type TLDrawShape,
	type TLDrawShapeProps,
	type TLDrawShapeSegment,
} from './shapes/TLDrawShape'
export {
	EMBED_DEFINITIONS,
	embedShapeTypeMigrations,
	embedShapeTypeValidator,
	tlEmbedShapePermissionDefaults,
	type EmbedDefinition,
	type TLEmbedShape,
	type TLEmbedShapePermissionName,
	type TLEmbedShapePermissions,
	type TLEmbedShapeProps,
} from './shapes/TLEmbedShape'
export {
	frameShapeTypeMigrations,
	frameShapeTypeValidator,
	type TLFrameShape,
	type TLFrameShapeProps,
} from './shapes/TLFrameShape'
export {
	geoShapeTypeMigrations,
	geoShapeTypeValidator,
	type TLGeoShape,
	type TLGeoShapeProps,
} from './shapes/TLGeoShape'
export {
	groupShapeTypeMigrations,
	groupShapeTypeValidator,
	type TLGroupShape,
	type TLGroupShapeProps,
} from './shapes/TLGroupShape'
export {
	highlightShapeTypeMigrations,
	highlightShapeTypeValidator,
	type TLHighlightShape,
	type TLHighlightShapeProps,
} from './shapes/TLHighlightShape'
export {
	iconShapeTypeMigrations,
	iconShapeTypeValidator,
	type TLIconShape,
	type TLIconShapeProps,
} from './shapes/TLIconShape'
export {
	imageShapeTypeMigrations,
	imageShapeTypeValidator,
	type TLImageCrop,
	type TLImageShape,
	type TLImageShapeProps,
} from './shapes/TLImageShape'
export {
	lineShapeTypeMigrations,
	lineShapeTypeValidator,
	type TLLineShape,
	type TLLineShapeProps,
} from './shapes/TLLineShape'
export {
	noteShapeTypeMigrations,
	noteShapeTypeValidator,
	type TLNoteShape,
	type TLNoteShapeProps,
} from './shapes/TLNoteShape'
export {
	textShapeTypeMigrations,
	textShapeTypeValidator,
	type TLTextShape,
	type TLTextShapeProps,
} from './shapes/TLTextShape'
export {
	videoShapeTypeMigrations,
	videoShapeTypeValidator,
	type TLVideoShape,
	type TLVideoShapeProps,
} from './shapes/TLVideoShape'
export { storeMigrations } from './store-migrations'
export { TL_ALIGN_TYPES, alignValidator, type TLAlignStyle, type TLAlignType } from './styles/align'
export {
	TL_ARROWHEAD_TYPES,
	arrowheadValidator,
	type TLArrowheadEndStyle,
	type TLArrowheadStartStyle,
	type TLArrowheadType,
} from './styles/arrowhead'
export { TL_STYLE_TYPES, type TLBaseStyle, type TLStyleType } from './styles/base-style'
export { colorValidator, type TLColorStyle, type TLColorType } from './styles/color'
export { TL_DASH_TYPES, dashValidator, type TLDashStyle, type TLDashType } from './styles/dash'
export { TL_FILL_TYPES, fillValidator, type TLFillStyle, type TLFillType } from './styles/fill'
export { TL_FONT_TYPES, fontValidator, type TLFontStyle, type TLFontType } from './styles/font'
export { TL_GEO_TYPES, geoValidator, type TLGeoStyle, type TLGeoType } from './styles/geo'
export { TL_ICON_TYPES, iconValidator, type TLIconStyle, type TLIconType } from './styles/icon'
export {
	TL_OPACITY_TYPES,
	opacityValidator,
	type TLOpacityStyle,
	type TLOpacityType,
} from './styles/opacity'
export { TL_SIZE_TYPES, sizeValidator, type TLSizeStyle, type TLSizeType } from './styles/size'
export {
	TL_SPLINE_TYPES,
	splineValidator,
	type TLSplineType,
	type TLSplineTypeStyle,
} from './styles/spline'
export { type TLStyleCollections, type TLStyleItem, type TLStyleProps } from './styles/style-types'
export { type TLVerticalAlignType } from './styles/vertical-align'
export { getDefaultTranslationLocale } from './translations/translations'
export { type SetValue, type SmooshedUnionObject } from './util-types'

// TL_ALIGN_TYPES
// TL_COLOR_TYPES
// TL_DASH_TYPES
// TL_FILL_TYPES
// TL_FONT_TYPES
// TL_OPACITY_TYPES
// TL_SIZE_TYPES
// TL_STYLE_TYPES
// alignValidator
// colorValidator
// dashValidator
// fillValidator
// fontValidator
// geoValidator
// opacityValidator
// sizeValidator
