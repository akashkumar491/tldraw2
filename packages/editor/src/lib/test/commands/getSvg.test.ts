import { createCustomShapeId } from '@tldraw/tlschema'
import { SVG_PADDING } from '../../constants'
import { TestApp } from '../TestEditor'

let app: TestApp

const ids = {
	boxA: createCustomShapeId('boxA'),
	boxB: createCustomShapeId('boxB'),
	boxC: createCustomShapeId('boxC'),
}

beforeEach(() => {
	app = new TestApp()
	app.setProp('dash', 'solid')
	app.createShapes([
		{
			id: ids.boxA,
			type: 'geo',
			x: 0,
			y: 0,
			props: {
				w: 100,
				h: 100,
				text: 'Hello world',
			},
		},
		{
			id: ids.boxB,
			type: 'geo',
			x: 100,
			y: 100,
			props: {
				w: 50,
				h: 50,
			},
		},
		{
			id: ids.boxC,
			type: 'geo',
			x: 400,
			y: 400,
			props: {
				w: 100,
				h: 100,
			},
		},
	])
	app.selectAll()
})

it('gets an SVG', async () => {
	const svg = await app.getSvg(app.selectedIds)

	expect(svg).toBeTruthy()
})

it('Does not get an SVG when no ids are provided', async () => {
	const svg = await app.getSvg([])

	expect(svg).toBeFalsy()
})

it('Gets the bounding box at the correct size', async () => {
	const svg = await app.getSvg(app.selectedIds)
	const bbox = app.selectionBounds!
	const expanded = bbox.expandBy(SVG_PADDING) // adds 32px padding

	expect(svg!.getAttribute('width')).toMatch(expanded.width + '')
	expect(svg!.getAttribute('height')).toMatch(expanded.height + '')
})

it('Gets the bounding box at the correct size', async () => {
	const svg = (await app.getSvg(app.selectedIds))!
	const bbox = app.selectionBounds!
	const expanded = bbox.expandBy(SVG_PADDING) // adds 32px padding

	expect(svg!.getAttribute('width')).toMatch(expanded.width + '')
	expect(svg!.getAttribute('height')).toMatch(expanded.height + '')
})

it('Matches a snapshot', async () => {
	const svg = (await app.getSvg(app.selectedIds))!

	const elm = document.createElement('wrapper')
	elm.appendChild(svg)

	expect(elm).toMatchSnapshot('Basic SVG')
})

it('Accepts a scale option', async () => {
	const svg1 = (await app.getSvg(app.selectedIds, { scale: 1 }))!

	expect(svg1.getAttribute('width')).toBe('564')

	const svg2 = (await app.getSvg(app.selectedIds, { scale: 2 }))!

	expect(svg2.getAttribute('width')).toBe('1128')
})

it('Accepts a background option', async () => {
	const svg1 = (await app.getSvg(app.selectedIds, { background: true }))!

	expect(svg1.style.backgroundColor).not.toBe('transparent')

	const svg2 = (await app.getSvg(app.selectedIds, { background: false }))!

	expect(svg2.style.backgroundColor).toBe('transparent')
})
