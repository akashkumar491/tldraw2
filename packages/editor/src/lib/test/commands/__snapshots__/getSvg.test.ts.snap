// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches a snapshot: Basic SVG 1`] = `
<wrapper>
  <svg
    direction="ltr"
    height="564"
    stroke-linecap="round"
    stroke-linejoin="round"
    style="background-color: transparent;"
    viewBox="-32 -32 564 564"
    width="564"
  >
    <defs>
      <mask
        id="hash_pattern_mask"
      >
        
					
        <rect
          fill="white"
          height="8"
          width="8"
          x="0"
          y="0"
        />
        
					
        <g
          stroke="black"
          strokelinecap="round"
        >
          
						
          <line
            x1="0.6666666666666666"
            x2="2"
            y1="2"
            y2="0.6666666666666666"
          />
          
						
          <line
            x1="3.333333333333333"
            x2="4.666666666666666"
            y1="4.666666666666666"
            y2="3.333333333333333"
          />
          
						
          <line
            x1="6"
            x2="7.333333333333333"
            y1="7.333333333333333"
            y2="6"
          />
          
					
        </g>
        
				
      </mask>
      <pattern
        height="8"
        id="hash_pattern"
        patternUnits="userSpaceOnUse"
        width="8"
      >
        
					
        <rect
          fill=""
          height="8"
          mask="url(#hash_pattern_mask)"
          width="8"
          x="0"
          y="0"
        />
        
				
      </pattern>
      <style />
    </defs>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 0, 0)"
    >
      <g>
        <path
          d="M0, 0L100, 0,100, 100,0, 100Z"
          fill="none"
          stroke=""
          stroke-width="3.5"
        />
        <g>
          <text
            alignment-baseline="mathematical"
            dominant-baseline="mathematical"
            fill=""
            font-family=""
            font-size="22px"
            font-style="normal"
            font-weight="normal"
            line-height="29.700000000000003px"
            stroke=""
            stroke-width="2"
          >
            <tspan
              alignment-baseline="mathematical"
              x="16px"
              y="-181px"
            >
              Hello world
            </tspan>
          </text>
          <text
            alignment-baseline="mathematical"
            dominant-baseline="mathematical"
            fill=""
            font-family=""
            font-size="22px"
            font-style="normal"
            font-weight="normal"
            line-height="29.700000000000003px"
            stroke="none"
          >
            <tspan
              alignment-baseline="mathematical"
              x="16px"
              y="-181px"
            >
              Hello world
            </tspan>
          </text>
        </g>
      </g>
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 100, 100)"
    >
      <path
        d="M0, 0L50, 0,50, 50,0, 50Z"
        fill="none"
        stroke=""
        stroke-width="3.5"
      />
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 400, 400)"
    >
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="none"
        stroke=""
        stroke-width="3.5"
      />
    </g>
  </svg>
</wrapper>
`;
